python/cudf/udf_cpp/shim.cu

```diff
@@ -662,3 +662,47 @@ make_definition_idx(BlockIdxMax, int64, int64_t);
 make_definition_idx(BlockIdxMax, float64, double);
 #undef make_definition_idx
 }
+
+extern "C" __device__ int BlockCorr(double* numba_return_value,
+                                    int64_t* const lhs_ptr,
+                                    int64_t* rhs_ptr,
+                                    int64_t size)
+{
+  double lhs_mean = BlockMean(lhs_ptr, size);
+  double rhs_mean = BlockMean(rhs_ptr, size);
+
+  // cuda::atomic<double, cuda::thread_scope_block> numerator = 0;
+  // cuda::atomic<double, cuda::thread_scope_block> sum_sq_l = 0;
+  // cuda::atomic<double, cuda::thread_scope_block> sum_sq_r = 0;
+
+  __shared__ double numerators[1024];
+  __shared__ double sum_sq_ls[1024];
+  __shared__ double sum_sq_rs[1024];
+
+  numerators[threadIdx.x] = 0.0;
+  sum_sq_ls[threadIdx.x]  = 0.0;
+  sum_sq_rs[threadIdx.x]  = 0.0;
+
+  auto block = cooperative_groups::this_thread_block();
+
+  for (int64_t idx = block.thread_rank(); idx < size; idx += block.size()) {
+    double delta_l = lhs_ptr[idx] - lhs_mean;
+    double delta_r = rhs_ptr[idx] - rhs_mean;
+
+    numerators[idx] = delta_l * delta_r;
```

Comment:
We should reimplement BlockVar to be a special case of a new function for computing Covariance (where the variance is the covariance of the data with itself). Then I think you can use the covariance function to compute the numerators and the variance function for the terms in the denominator.